key-concepts/covenants.adoc

```diff
@@ -1,11 +1,174 @@
+
+image:img/covenant-chains.svg["Each node in an Interbit blockchain
+network applies the same covenant", 500, 170, role="zoom center"]
+
+The covenant dictates how the blockchain can be changed over time. As a
+result, all participating nodes in the blockchain's network process
+actions and update the blockchain's state in a deterministic fashion.
+
+
+== Anatomy
+
+image:img/covenant-anatomy.svg["The anatomy of an Interbit covenant",
+522, 220, align="center" role="zoom center"]
+
+An Interbit covenant is comprised of several components:
+
+. A set of link:actions.adoc[action] creators and types. Any changes to
+  be made to a blockchain begin with an action creator that creates an
+  action of a type known to the blockchain.
+
+. A link[smart_contracts.adoc[smart contract]. An Interbit smart
+  contract is a block of Javascript code that processes each action
+  dispatched to the blockchain.
+
+. A link:sagas.adoc[saga]. A saga is used to coordinate any
+  non-deterministic operations such that only deterministic changes are
+  made to the blockchain.
+
+
+== Functional paradigm
+
+Interbit blockchains, and their covenants, adopt the functional
+programming paradigm. Functional programming (FP) is the process of
+writing software that uses pure functions, and avoids shared state, side
+effects, and mutable data.
+
+Pure functions have several important properties:
+
+. Given the same input, they always return the same output.
+
+. Executing a pure function produces no side effects.
+
+. The call to a pure function can be replaced by its resulting value
+  without changing the meaning of the program.
+
+It may be helpful to contrast pure functions with non-pure functions, so
+that you can see how useful these properties can be:
+
+[cols="1a,1a",options="header"]
+|===
+| Pure function
+| Non-pure function
+
+|
+[source,js]
+----
+x = 2
+y = 3
+function add(number1, number2) {
+  result = number1 + number2
+  return result
+}
+
+console.log(add(x, y))
+console.log(add(x, y))
+console.log(add(2, 3))
+console.log(add(x, y))
+console.log(x, y)
+----
+
+Outputs:
+[source]
+5
+5
+5
+5
+2 3
+
+|
+[source,js]
+----
+x = 2
+y = 3
+function add(number1, number2) {
+  result = x = x + number1 + number2
+  return result
+}
+
+console.log(add(x, y))
+console.log(add(x, y))
+console.log(add(2, 3))
+console.log(add(x, y))
+console.log(x, y)
+----
+
+Outputs:
+[source]
+7
+17
+22
+47
+47 3
+
+|===
+
+These are contrived examples, but as you can see:
+
+- The pure function produces the same result each time, and `x` and `y`
+  are unmodified.
+
+- The non-pure function produces different results each time, even when
+  explicitly called with the initial inputs, and the value of `x` has
+  changed notably (a side effect of calling the function).
+
+
+== Covenants and blockchains
+
+An Interbit blockchain applies a single covenant. Therefore, the
+covenant fully controls how the blockchain's state can be modified.
```

Comment:
~fully~

There are system actions that the blockchain itself performs which also modify state.